### PR TITLE
Refactor TreeSummarize

### PR DESCRIPTION
**Pull request by Disiok**
# Description

Refactor `TreeSummarize` with much simpler logic. Also eliminates the bug where we truncate text when < 10 nodes.

New logic is as simple as it gets:
1. repack text chunks (into maximal chunks that fills the context window)
2. if there's one chunk, summarize it and return as final response
3. otherwise, summarize all chunks in parallel, and recurse to step 1.

Benefits
* no more refine (hard to reason about), now just a basic recursive map / reduce (via repack)
* no more dependency on tree index / builder, easier path to deprecating legacy tree code

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: llama_index/indices/response/tree_summarize.py [hunk 1]
```diff
@@ -1,146 +1,148 @@
...
+        self._text_qa_template = text_qa_template or DEFAULT_TEXT_QA_PROMPT
         self._use_async = use_async
+        self._verbose = verbose
 
     @llm_token_counter("aget_response")
     async def aget_response(
         self,
         query_str: str,
         text_chunks: Sequence[str],
         prev_response: Optional[str] = None,
```
Comment (Disiok) [inline llama_index/indices/response/tree_summarize.py:36]: future PR: planning on cleaning up this interface and removing this `prev_response` thing

File: llama_index/indices/response/tree_summarize.py [hunk 2]
```diff
@@ -1,146 +1,148 @@
...
 
-class TreeSummarize(Refine):
+class TreeSummarize(BaseResponseBuilder):
     def __init__(
         self,
-        service_context: ServiceContext,
-        text_qa_template: QuestionAnswerPrompt,
-        refine_template: RefinePrompt,
+        service_context: Optional[ServiceContext] = None,
+        text_qa_template: Optional[QuestionAnswerPrompt] = None,
```
Comment (Disiok) [inline llama_index/indices/response/tree_summarize.py:18]: just making this easier to use on its own, prepping for future.

File: llama_index/indices/response/tree_summarize.py [hunk 3]
```diff
@@ -1,146 +1,148 @@
...
-    SummaryPrompt,
-)
-from llama_index.storage.docstore.registry import get_default_docstore
+from llama_index.prompts.prompts import QuestionAnswerPrompt, SummaryPrompt
 from llama_index.token_counter.token_counter import llm_token_counter
 from llama_index.types import RESPONSE_TEXT_TYPE
 
 
-class TreeSummarize(Refine):
+class TreeSummarize(BaseResponseBuilder):
```
Comment (Disiok) [inline llama_index/indices/response/tree_summarize.py:14]: TODO: add doc string to explain how this works.

File: tests/indices/response/test_response_builder.py [hunk 4]
```diff
@@ -105,50 +105,49 @@ def test_compact_response(mock_service_context: ServiceContext) -> None:
...
-        text_qa_template=mock_qa_prompt,
-        refine_template=mock_refine_prompt,
-    )
-
-    response = builder.get_response(
-        text_chunks=texts, query_str=query_str, num_children=2
-    )
-    # TODO: fix this output, the \n join appends unnecessary results at the end
-    assert str(response) == "What is?:This:is:a:bar:This:is:another:test"
+# TODO: fix test, right now results in infinite loop since mock LLM does not
```
Comment (Disiok) [inline tests/indices/response/test_response_builder.py:108]: TODO: remove or improve testing.

File: llama_index/indices/response/tree_summarize.py
Comment (logan-markewich) [inline llama_index/indices/response/tree_summarize.py:1]: This file is 100x easier to understand now. I can see whats going on even just by looking at the noisy git diff lol

**Review by logan-markewich — APPROVED** [review]
Looks good so far! I think one quick unit test for this should be enough

File: tests/indices/response/test_tree_summarize.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    query_str = "What is?"
+    texts = [
+        "Text chunk 1",
+        "Text chunk 2",
+        "Text chunk 3",
+        "Text chunk 4",
+    ]
+
+    response = tree_summarize.get_response(text_chunks=texts, query_str=query_str)
```
Comment (jerryjliu) [inline tests/indices/response/test_tree_summarize.py:41]: might as well try async too

**Review by jerryjliu — APPROVED** [review]
lgtm! way cleaner, thanks 

we're not really exposing tree summarize in docs but if once we make more of a push to make responsesynthesizer/responsebuidler modules user facing would be good to explicitly document

File: tests/indices/response/test_tree_summarize.py (hunk 5 above)
Comment (Disiok) [reply tests/indices/response/test_tree_summarize.py:72]: added test for async and `use_sync` versions.

**Pull request merged**